system-configuration/src/network_configuration.rs

```diff
@@ -0,0 +1,439 @@
+                        match router_str.parse::<IpAddr>() {
+                            Ok(router_ip) => {
+                                return Some(router_ip);
+                            }
+                            _ => { }
+                        }
+                    }
+                }
+            }
+        }
+
+        return None;
+    }
+
+    // pub fn netinfo(&self);
+    // pub fn proxies(&self) ;
+}
+
+
+pub struct SCNetworkService(pub SCNetworkServiceRef);
+
+impl SCNetworkService {
+    pub fn list() -> Vec<SCNetworkService> {
+        let prefs = unsafe {
+            SCPreferencesCreate(kCFAllocatorDefault,
+                                CFString::from_static_string("ns_list").as_concrete_TypeRef(),
+                                ptr::null())
+        };
+
+        let array: CFArray<SCNetworkServiceRef> = unsafe { 
+            CFArray::wrap_under_get_rule(SCNetworkServiceCopyAll(prefs)) 
+        };
+
+        array.get_all_values()
+              .iter()
+              .map(|service_ptr| SCNetworkService( unsafe { mem::transmute(*service_ptr) } ))
+              .collect::<Vec<SCNetworkService>>()
+    }
+    
+    pub fn list_order() -> Vec<SCNetworkService> {
+        let prefs = unsafe {
+            SCPreferencesCreate(kCFAllocatorDefault, 
+                                CFString::from_static_string("ns_list_order").as_concrete_TypeRef(),
+                                ptr::null())
+        };
+
+        let netset = unsafe { SCNetworkSetCopyCurrent(prefs) };
+
+        let array: CFArray<SCNetworkServiceRef> = unsafe {
+            CFArray::wrap_under_get_rule( SCNetworkSetGetServiceOrder(netset) )
+        };
+
+        let mut services = Vec::new();
+
+        for id in array.get_all_values().iter() {
+            let id_ptr: CFStringRef = unsafe { mem::transmute(*id) };
+            let service_ptr: SCNetworkServiceRef = unsafe { SCNetworkServiceCopy(prefs, id_ptr) };
+            services.push(SCNetworkService(service_ptr));
+        }
+
+        services
+    }
+
+    pub fn id(&self) -> String {
+        unsafe { CFString::wrap_under_get_rule( SCNetworkServiceGetServiceID( self.0 ) ) }.to_string()
+    }
+
+    pub fn name(&self) -> String {
+        unsafe { CFString::wrap_under_get_rule( SCNetworkServiceGetName( self.0 ) ) }.to_string()
+    }
+
+    pub fn enabled(&self) -> bool {
+        let ret = unsafe { SCNetworkServiceGetEnabled( self.0 ) };
+        ret == 1
+    }
+
+    pub fn dns(&self) -> SCNetworkServiceDNS {
+        let store = SCDynamicStoreBuilder::new("ns_dns").build();
+
+        let mut state_domain_name: Option<String> = None;
+        let mut state_server_addresses: Option<Vec<IpAddr>> = None;
+        let mut setup_domain_name: Option<String> = None;
+        let mut setup_server_addresses: Option<Vec<IpAddr>> = None;
+
+        if let Some(value) = store.get(CFString::new(&format!("State:/Network/Service/{}/DNS", self.id()))) {
```

Comment:
If State and Setup branches do the same perhaps this can be turned into a loop or maybe a helper function.

Reply:
Yeah.